### PR TITLE
Use a distinct lock id for Istiod

### PR DESCRIPTION
**Pull request by howardjohn**
Currently leader election is used for

* config map syncer, as an optimization
* webhook controller, as an optimization
* ingress status, not sure if this is needed for correctness or
optimization

The problem that arises is when an older istiod that only has some
subset of this functionality (for example, legacy helm install will only
have ingress syncer) holds the lock, preventing the new Istiod from
running properly (as the config map syncer does not exist). In order to
get around this, this PR proposes changing the lock name. This is a
pretty poor workaround and will break the next time we add new
functionality - I am very open to new ideas here

For https://github.com/istio/istio/issues/22874

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
Not sure i understand what you said

**Comment by howardjohn** [comment]
The old control plane will hold the lock. The old control plane does not
have all the same functionality, so it's holding the lock but not
performing all of the functionality required (the config map controller
specifically)

**Comment by hzxuzhonghu** [comment]
But when user rolling update, doesn't the old control plane terminated, and the new one runs up and holds the lock?

**Comment by howardjohn** [comment]
This is for multiple control plane where they both run at the same time

**Comment by hzxuzhonghu** [comment]
One cluster multiple control plane? How could this solve it?

**Comment by howardjohn** [comment]
Yes we support multiple control planes now. See
https://preliminary.istio.io/blog/2020/multiple-control-planes/

**Comment by costinm** [comment]
My take: long term we may want to designate a revision as 'primary' - handling the 'singleton' functions, possibly acting as a proxy for other revisions for injection. 
The 'primary' could also patch the webhooks and be the only one with broader RBAC permissions.

In a single-control plane mode - this behavior is identical with what we have in 1.5 and for in-place upgrades. In multi-revision: we designate the primary trough an option or using the un-revisioned name ( istiod.istio-systm ) - and we can drop RBAC and features from non-primary. 

The real problem is that the master election must be across revisions - if we go with the 'primary' we have a deterministic master.

**Comment by hzxuzhonghu** [comment]
@howardjohn I can think of that for different components, make them use different leader electors. So there will be no such issue.

**Comment by howardjohn** [comment]
Yeah that was my other idea but I am not sure if it's too complicated. I am
ok with either - want me to switch to that approach?

**Comment by hzxuzhonghu** [comment]
Yeah, personally i prefer that.

**Comment by howardjohn** [comment]
@hzxuzhonghu I switched to multiple different locks

File: pilot/pkg/bootstrap/server.go [hunk 1]
```diff
@@ -291,24 +280,30 @@ func NewServer(args *PilotArgs) (*Server, error) {
...
+				return map[string]string{
+					constants.CACertNamespaceConfigMapDataName: string(s.ca.GetCAKeyCertBundle().GetRootCertPem()),
+				}
+			}
+			s.addTerminatingStartFunc(func(stop <-chan struct{}) error {
+				leaderelection.
+					NewLeaderElection(args.Namespace, args.PodName, leaderelection.NamespaceController, s.kubeClient).
+					AddRunFunction(func(stop <-chan struct{}) {
+						log.Infof("Starting namespace controller")
+						nc := NewNamespaceController(fetchData, args.Config.ControllerOptions, s.kubeClient)
```
Comment (hzxuzhonghu) [inline pilot/pkg/bootstrap/server.go:300]: why move namespace controller here

**Review by hzxuzhonghu — APPROVED** [review]

File: pilot/pkg/bootstrap/server.go (hunk 1 above)
Comment (howardjohn) [reply pilot/pkg/bootstrap/server.go:300]: before it's nesting 2 addStartFunc calls which ends up behaving weird so I broke it out

**Pull request merged**